File: my04-Maze-Solver迷宫解密/maze.cpp
#include<iostream>
#include<cstdlib>
#include<opencv2/core/core.hpp>
#include<opencv2/highgui/highgui.hpp>
#include<opencv2/imgproc/imgproc.hpp>

using namespace std;
using namespace cv;
Mat kernel = Mat::ones(15, 15, CV_8UC1);
class Morph{
private:
	int dilationElem,erodeElem;
	int dilationSize,erodeSize;
	
public:
	Morph(){
		dilationElem=0;
		erodeElem=0;
		dilationSize=2;
		erodeSize=2;
	}
	Mat dilateImage(Mat input){
		Mat temp,element;
		int dilationType;
		if(dilationElem==0)
			dilationType=MORPH_RECT;
		else if(dilationElem==1)
			dilationType=MORPH_CROSS;
		else if(dilationElem==2)
			dilationType=MORPH_ELLIPSE;
		element= getStructuringElement(dilationType,Size(2*dilationSize+1,2*dilationSize+1),Point(dilationSize,dilationSize));
		dilate(input,temp,kernel);
		return temp;
	}
	Mat erodeImage(Mat input){
		Mat temp,element;
		int erodeType;
		if(erodeElem==0)
			erodeType=MORPH_RECT;
		else if(erodeElem==1)
			erodeType=MORPH_CROSS;
		else if(erodeElem==2)
			erodeType=MORPH_ELLIPSE;
		element= getStructuringElement(erodeType,Size(2*erodeSize+1,2*erodeSize+1),Point(erodeSize,erodeSize));
		dilate(input,temp,kernel);
		return temp;
	}
};
int main(int argc, char **argv){
	if(argc!=2){
		cout<<"Wait for an image"<<endl;
		return -1;
	}
	vector<vector<Point> > contours;
	Mat inputMaze,gray,binary,dilation,erosion,imgDiff,BGRcomp[3],imgDiff_inv,output,red,green;
	Morph mp;
	inputMaze=imread(argv[1],CV_LOAD_IMAGE_COLOR);
	cvtColor(inputMaze,gray,CV_BGR2GRAY);
	threshold(gray,binary,127,255,CV_THRESH_BINARY_INV);

	findContours(binary,contours,CV_RETR_EXTERNAL, CV_CHAIN_APPROX_NONE);
	drawContours(binary, contours, 0, CV_RGB(255,255,255), CV_FILLED);
	threshold(binary,binary,240,255,CV_THRESH_BINARY);

	dilation=mp.dilateImage(binary);
	erosion=mp.erodeImage(dilation);
	absdiff(dilation,erosion,imgDiff);
	bitwise_not(imgDiff,imgDiff_inv);
	split(inputMaze,BGRcomp);
	namedWindow("diff",WINDOW_AUTOSIZE);
	imshow("diff",imgDiff);
	bitwise_and(BGRcomp[2],BGRcomp[2],red,imgDiff_inv);
	bitwise_and(BGRcomp[1],BGRcomp[1],green,imgDiff_inv);
	BGRcomp[2]=red.clone();
	BGRcomp[1]=green.clone();
	merge(BGRcomp,3,output);
	namedWindow("SolvedMaze",WINDOW_AUTOSIZE);
	imshow("SolvedMaze",output);
	//imwrite("OutputMaze.jpg",output);
	waitKey(0);
	return 0;
}

File: my04-Maze-Solver迷宫解密/maze.py


'''
源文件是使用opencv2.4，
改成opencv3.2有点问题。
https://ishankgulati.github.io/posts/Maze-Solver/
'''
import cv2
import numpy as np



img = cv2.imread('SampleImages/1.png')
# img = cv2.imread('SampleImages/2.png')
# img = cv2.imread('SampleImages/3.jpg')#不行，得修改
# img = cv2.imread('SampleImages/huge_maze.jpg')#不行，得修改
cv2.imshow('maze',img)
cv2.waitKey(0)

# Binary conversion
gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
ret, thresh = cv2.threshold(gray, 127, 255, cv2.THRESH_BINARY_INV)#反转tholdolding将给我们一个二进制的图像与白色的墙壁和黑色的背景。
cv2.imshow('THRESH_BINARY_INV',thresh)
cv2.waitKey(0)

# Contours
image,contours, hierarchy = cv2.findContours(thresh, cv2.RETR_EXTERNAL,
                                       cv2.CHAIN_APPROX_NONE)
print('len(contours):',len(contours))
# dc=cv2.drawContours(thresh, contours, 0, (255, 255, 255), -1)
dc=cv2.drawContours(thresh, contours, 0, (255, 255, 255), 5)#用不同颜色来标注
dc=cv2.drawContours(dc, contours, 1, (0, 0, 0), 5)# TODO 大迷宫的len(contours): 26
cv2.imshow('drawContours',dc)
cv2.waitKey(0)

ret, thresh = cv2.threshold(dc, 240, 255, cv2.THRESH_BINARY)
# ret, thresh = cv2.threshold(thresh, 240, 255, cv2.THRESH_BINARY)
cv2.imshow('thresh2',thresh)
cv2.waitKey(0)

# Dilate
'''
扩张

扩张是数学形态领域的两个基本操作者之一，另一个是侵蚀。它通常应用于二进制图像，但有一些版本可用于灰度图像。操作者对二进制图像的基本效果是逐渐扩大前景像素区域的边界（通常为白色像素）。因此，前景像素的面积大小增加，而这些区域内的孔变小。
'''
ke = 10
# kernel = np.ones((19, 19), np.uint8)
kernel = np.ones((ke, ke), np.uint8)
dilation = cv2.dilate(thresh, kernel, iterations=1)
cv2.imshow('dilation',dilation)
cv2.waitKey(0)

# Erosion
#侵蚀是第二个形态运算符。它也适用于二进制图像。操作者对二进制图像的基本效果是消除前景像素区域的边界（通常为白色像素）。因此，前景像素的面积缩小，并且这些区域内的孔变大。
erosion = cv2.erode(dilation, kernel, iterations=1)
cv2.imshow('erosion',erosion)
cv2.waitKey(0)

#找到两个图像的差异
diff = cv2.absdiff(dilation, erosion)
cv2.imshow('diff',diff)
cv2.waitKey(0)

# splitting the channels of maze
b, g, r = cv2.split(img)
mask_inv = cv2.bitwise_not(diff)
#为了在原始迷宫图像上显示解决方案，首先将原来的迷宫分割成r，g，b组件。现在通过反转diff图像创建一个掩码。使用在最后一步中创建的掩码的原始迷宫的按位和r和g分量。这一步将从迷宫解决方案的图像部分去除红色和绿色成分。最后一个是合并所有组件，我们将使用蓝色标记的解决方案。
# masking out the green and red colour from the solved path
r = cv2.bitwise_and(r, r, mask=mask_inv)
g = cv2.bitwise_and(g, g, mask=mask_inv)

res = cv2.merge((b, g, r))
cv2.imshow('Solved Maze', res)
cv2.imwrite('SampleImages/Solved-Maze-1.png',res)

cv2.waitKey(0)
cv2.destroyAllWindows()
